internal/controller/pkg/resolver/reconciler.go

```diff
@@ -375,3 +401,95 @@ func (r *Reconciler) findDependencyVersion(ctx context.Context, dep *v1beta1.Dep
 
 	return addVer, nil
 }
+
+// UpdatableVersionFinder is an implementation of versionFinder that allows for version upgrade/downgrade capability considering parent constraints.
+type UpdatableVersionFinder struct {
+	fetcher xpkg.Fetcher
+	config  xpkg.ConfigStore
+}
+
+// FindValidDependencyVersion finds a valid version with version upgrade/downgrade capability considering parent constraints.
+func (u *UpdatableVersionFinder) FindValidDependencyVersion(ctx context.Context, dep *v1beta1.Dependency, ref name.Reference, n dag.Node, log logging.Logger) (string, error) {
+	if digest, err := conregv1.NewHash(dep.Constraints); err == nil {
+		log.Debug("package is pinned to a specific digest, skipping resolution")
+		return digest.String(), nil
+	}
+
+	ic, ps, err := u.config.PullSecretFor(ctx, ref.String())
+	if err != nil {
+		log.Info("cannot get pull secret from image config store", "error", err)
+		return "", errors.Wrap(err, errGetPullConfig)
+	}
+
+	var s []string
+	if ps != "" {
+		log.Debug("Selected pull secret from image config store", "image", ref.String(), "imageConfig", ic, "pullSecret", ps)
+		s = append(s, ps)
+	}
+	// NOTE(hasheddan): we will be unable to fetch tags for private
+	// dependencies because we do not attach any secrets. Consider copying
+	// secrets from parent dependencies.
+	tags, err := u.fetcher.Tags(ctx, ref, s...)
+	if err != nil {
+		log.Debug(errFetchTags, "error", err)
+		return "", errors.New(errFetchTags)
+	}
+
+	vs := []*semver.Version{}
+	for _, r := range tags {
+		v, err := semver.NewVersion(r)
+		if err != nil {
+			// We skip any tags that are not valid semantic versions.
+			continue
+		}
+		vs = append(vs, v)
+	}
+
+	var cv *semver.Version
+
+	// If the node is a LockPackage, it means it's already installed.
+	if lp, ok := n.(*v1beta1.LockPackage); ok {
+		log.Debug("dependency found in lock, trying to find the minimum valid version which is greater than the current version")
+		sort.Sort(semver.Collection(vs))
+		cv = semver.MustParse(lp.Version)
+	}
+
+	// If the node is not a LockPackage, it means it's a new dependency that we need to install with the latest version.
+	if cv == nil {
+		log.Debug("dependency not found in lock, trying to find the maximum valid version")
+		sort.Sort(sort.Reverse(semver.Collection(vs)))
```

Comment:
I know why we are doing this, but for posterity, it would be helpful to add a small comment on why we reverse sort here vs regular sort above 🙏